src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+    drive->silent++;
+    return 0;
+  }
+
+  if (!drive->silent)
+    printf("** Testing DMA speed...\n");
+  st = getmsecs();
+  for (et = getmsecs(); (et - st) < (msecs); et = getmsecs()) {
+    if (write_buffer(drive, 0x0A, 0, 0, buf_size)) {
+      if (!drive->silent)
+        printf("WRITE BUFFER error! DMA speed test aborted!\n");
+      drive->silent++;
+      return 1;
+    }
+    if (read_buffer(drive, 0x0A, 0, 0, buf_size)) {
+      if (!drive->silent)
+        printf("READ BUFFER error! DMA speed test aborted!\n");
+      drive->silent++;
+      return 1;
+    }
+    cnt++;
+  }
+  speed = (buf_size * cnt / msecs);
+  printf("DMA speed: %6ld kB/s (%d buffers of %d bytes in %ld msecs)\n", speed,
+         cnt, buf_size, msecs);
+  drive->silent++;
+  return 0;
+}
+
+int flush_cache(drive_info *drive, bool IMMED) {
+  //	drive->cmd[0] = SBC_FLUSH_CACHE;
+  drive->cmd[0] = MMC_SYNC_CACHE;
+  drive->cmd[1] = IMMED ? 0x02 : 0;
+
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    if (!drive->silent)
+      sperror("SBC_FLUSH_CACHE", drive->err);
+    return (drive->err);
+  }
+  //		{ sperror ("SBC_FLUSH_CACHE",drive->err); return (drive->err); }
+
+  return 0;
+}
+
+int set_cache(drive_info *drive, bool rd, bool wr) {
+  memset(drive->rd_buf, 0, 20);
+  drive->rd_buf[8] = MODE_PAGE_CACHING;
+  drive->rd_buf[9] = 0x0A;
+  drive->rd_buf[10] = (wr ? 0x04 : 0) | (rd ? 0 : 0x01);
+
+  return (mode_select(drive, 0x20));
+}
+
+int get_cache(drive_info *drive, bool *rd, bool *wr) {
+  bool re, we;
+  if (mode_sense(drive, MODE_PAGE_CACHING, 0, 20) ||
+      drive->rd_buf[8] != MODE_PAGE_CACHING)
+    return 1;
+  re = !(drive->rd_buf[10] & 0x01);
+  we = !!(drive->rd_buf[10] & 0x04);
+  printf("Cache:  RD %s, WR %s\n", re ? "EN" : "DIS", we ? "EN" : "DIS");
+  if (rd)
+    (*rd) = re;
+  if (wr)
+    (*wr) = we;
+  return 0;
+}
+
+int get_mode_pages_list(drive_info *drive) {
+  unsigned int len, i, ii;
+  unsigned char ml = 0, mn = 0;
+  if (!drive->silent)
+    printf("\n** Reading supported mode pages...\n");
+  if (mode_sense(drive, 0x3F, 2, 0x4000))
+    return 1;
+  len = ntoh16u(drive->rd_buf);
+  //	printf("data len: %4X (%4d), Header:\n", len, len);
+  //	for (i=0; i<8; i++) printf(" 0x%02X",drive->rd_buf[i] & 0xFF);
+  //printf("\n");
+  for (i = 8; (i < len) && (i < 0x4000); i += ml) {
+    mn = drive->rd_buf[i] & 0x3F;
+    ml = drive->rd_buf[i + 1] & 0xFF;
+    ii = 0;
+    while ((MODE_PAGES[ii].id != mn) && (MODE_PAGES[ii].id < 0x3F))
+      ii++;
+    if (!drive->silent) {
+      printf("Mode Page: 0x%02X [%s]", mn, MODE_PAGES[ii].name);
+#ifdef MODE_PAGES_DEBUG
+      for (ii = 0; ii < (ml + 2); ii++) {
+        if (!(ii % 32))
+          printf("\n");
+        printf(" %02X", drive->rd_buf[i + ii] & 0xFF);
+      }
+#endif
+      printf("\n");
+    }
+    ml += 2;
+  }
+  return 0;
+}
+
+int get_features_list(drive_info *drive) {
+  unsigned int len, i, ii;
+  unsigned short fn;
+  unsigned char fv;
```

Comment:
Codacy found an issue: [The scope of the variable 'fv' can be reduced.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)